core/src/outstanding_requests.rs

```diff
@@ -0,0 +1,166 @@
+use crate::request_response::RequestResponse;
+use lru::LruCache;
+use rand::{thread_rng, Rng};
+use solana_ledger::shred::Nonce;
+use solana_sdk::timing::timestamp;
+
+pub const DEFAULT_REQUEST_EXPIRATION_MS: u64 = 60_000;
+
+pub struct OutstandingRequests<T, S>
+where
+    T: RequestResponse<Response = S>,
+{
+    requests: LruCache<Nonce, RequestStatus<T>>,
+}
+
+impl<T, S> OutstandingRequests<T, S>
+where
+    T: RequestResponse<Response = S>,
+{
+    // Returns boolean indicating whether sufficient time has passed for a request with
+    // the given timestamp to be made
+    pub fn add_request(&mut self, request: T) -> Nonce {
+        let num_expected_responses = request.num_expected_responses();
+        let nonce = thread_rng().gen_range(0, Nonce::MAX);
+        self.requests.put(
+            nonce,
+            RequestStatus {
+                expire_timestamp: timestamp() + DEFAULT_REQUEST_EXPIRATION_MS,
+                num_expected_responses,
+                request,
+            },
+        );
+        nonce
+    }
+
+    pub fn register_response(&mut self, nonce: u32, response: &S, now: u64) -> bool {
+        let (is_valid, should_delete) = self
+            .requests
+            .get_mut(&nonce)
+            .map(|status| {
+                if status.num_expected_responses > 0
+                    && now < status.expire_timestamp
```

Comment:
shouldn't it also remove the key if `now > expire_timestamp`?

also this looks like it keeps the nonce alive as long as the sender is sending invalid responses because those will not decrease `num_expected_responses`. shouldn't invalid responses also count towards `num_expected_responses`?

Reply:
I think it does remove the key if `now > expire_timestamp` no? If `now < status.expire_timestamp` is false, then it will go into the `else` which will set `should_delete=true`.

See this test:
https://github.com/solana-labs/solana/pull/16512/files#diff-d0fb1321af935f265f21af26229db3ace2fda6083c96a53f1d4ef58d2deb5b76R94-R107

Reply:
> shouldn't invalid responses also count towards num_expected_responses

Maybe. I can't really think of an attack from it. Is there a downside?

Reply:
> I think it does remove the key if now > expire_timestamp no?

true, my mistake